### ISSUE TITLE
Pre-#182 checkpoint resume silently casts bf16 Adam state to fp32 (lossy)

### ISSUE THREAD

**Issue by shuheng-liu** [TensorAuto/OpenTau#184]
## Summary

Optimizer checkpoints saved before #182 lands hold bf16 `exp_avg` / `exp_avg_sq` (because the inner `torch.optim.AdamW` was constructed over bf16 params). After #182, the inner AdamW operates over fp32 master parameters, so its newly-allocated state is fp32. On resume from a pre-#182 checkpoint, `torch.optim.Optimizer.load_state_dict` auto-casts the loaded state tensors to match the current param dtype (now fp32) — so loads succeed, but the bf16 mantissa precision is what you get; the missing fp32 bits are gone for good.

This is **functional, not a crash**. Worth a note for anyone resuming a pi05 / libero run from a pre-#182 (or pre-#176, since DDP under #176 also produced bf16 state) checkpoint.

## Reproduction

```bash
# 1. Train for a few steps under DDP on a commit before #182 lands
#    -> checkpoint contains bf16 Adam state.
# 2. Pull main (with #182 merged) and resume from that checkpoint:
opentau-train --resume \
    --accelerate-config=configs/libero/reproduce_pi05_libero_accelerate_config.yaml \
    --config_path=configs/libero/reproduce_pi05_libero.json \
    --checkpoint_path=<old_checkpoint>

# Adam will run with state that was bf16-rounded at save-time and is now
# fp32-typed but bf16-truthful. No error, just slightly noisier moments
# until fresh fp32 state accumulates around them.
```

## What "lossy" means concretely

bf16 has ~7 bits of mantissa; fp32 has ~23. On save, every value in `exp_avg` / `exp_avg_sq` was rounded to ~7 bits. On load, those rounded values are re-typed as fp32, so the bottom ~16 bits are zero. Adam keeps running, and within a few hundred steps the moments fold in fresh fp32 gradients and the noise floor recovers. Bounded one-time degradation, not compounding.

## Options

1. **Document and accept** (current state). Cheapest. Note in the PR / release notes that resume from a pre-#182 checkpoint pays a one-time precision tax.
2. **Add a guard with a warning.** In `MasterWeightOptimizer.load_state_dict`, inspect the loaded state dtypes; if any tensor is bf16 (or fp16), log a `UserWarning` pointing to this issue and the workaround (start fresh, or re-train a few hundred steps to wash it out).
3. **Add an upgrade utility.** A one-shot script that loads a pre-#182 optimizer state and writes it back with bf16->fp32 promotion explicit. Doesn't recover precision (impossible — the bits are gone), but makes the conversion auditable.

I'd vote for 1 + 2 — the warning is ~5 lines and prevents silent surprise. Option 3 doesn't materially help.

## Owner / scope

* Touches: `src/opentau/optim/master_weights.py::load_state_dict`.
* No tests required for option 1; option 2 needs a one-line `pytest.warns(UserWarning)` test.

Refs: #181, #182 (introduces the wrapper).
